stablehlo/dialect/TypeInference.cpp

```diff
@@ -2391,10 +2391,24 @@ LogicalResult inferSliceOp(Optional<Location> location, Value operand,
 }
 
 LogicalResult inferSortOp(
-    Optional<Location>, ValueRange inputs,
+    Optional<Location> location, ValueRange inputs,
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  for (auto resultType : inputs.getTypes())
-    inferredReturnShapes.emplace_back(resultType.cast<ShapedType>());
+  // Infer most specific type from all operands
+  SmallVector<Type> mostSpecificType;
+  if (failed(hlo::inferMostSpecificType(location, inputs.getTypes(),
+                                        mostSpecificType)))
+    return failure();
+
+  auto rankedOutput = mostSpecificType.front().dyn_cast<RankedTensorType>();
+  for (auto inputType : inputs.getTypes()) {
+    auto inputShapedType = inputType.cast<ShapedType>();
+    if (rankedOutput)
+      inferredReturnShapes.emplace_back(rankedOutput.getShape(),
```

Comment:
add a comment mentioning that we preserve the original type and only use shape/bounds from the inferred result.